contracts/proxies/GnosisSafeProxyFactory.sol

```diff
@@ -54,16 +34,41 @@ contract GnosisSafeProxyFactory {
         require(address(proxy) != address(0), "Create2 call failed");
     }
 
-    /// @dev Allows to create new proxy contact and execute a message call to the new proxy within one transaction.
-    /// @param _singleton Address of singleton contract.
+    /// @dev Allows to create a new proxy contract and execute a message call to the new proxy within one transaction.
+    /// @param _singleton Address of singleton contract. Must be deployed at the time of execution.
     /// @param initializer Payload for message call sent to new proxy contract.
     /// @param saltNonce Nonce that will be used to generate the salt to calculate the address of the new proxy contract.
     function createProxyWithNonce(
         address _singleton,
         bytes memory initializer,
         uint256 saltNonce
     ) public returns (GnosisSafeProxy proxy) {
-        proxy = deployProxyWithNonce(_singleton, initializer, saltNonce);
+        // If the initializer changes the proxy address should change too. Hashing the initializer data is cheaper than just concatinating it
+        bytes32 salt = keccak256(abi.encodePacked(keccak256(initializer), saltNonce));
+        proxy = deployProxy(_singleton, salt);
+        if (initializer.length > 0)
+            // solhint-disable-next-line no-inline-assembly
+            assembly {
+                if eq(call(gas(), proxy, 0, add(initializer, 0x20), mload(initializer), 0, 0), 0) {
+                    revert(0, 0)
+                }
+            }
+        emit ProxyCreation(proxy, _singleton);
+    }
+
+    /// @dev Allows to create a new proxy contract that should exist only on 1 network (e.g. specific governance or admin accounts)
+    ///      by including the chain id in the create2 salt. Such proxies cannot be created on other networks by replaying the transaction.
+    /// @param _singleton Address of singleton contract. Must be deployed at the time of execution.
+    /// @param initializer Payload for message call sent to new proxy contract.
+    /// @param saltNonce Nonce that will be used to generate the salt to calculate the address of the new proxy contract.
+    function createChainSpecificProxyWithNonce(
+        address _singleton,
+        bytes memory initializer,
+        uint256 saltNonce
+    ) public returns (GnosisSafeProxy proxy) {
+        // If the initializer changes the proxy address should change too. Hashing the initializer data is cheaper than just concatinating it
+        bytes32 salt = keccak256(abi.encodePacked(keccak256(initializer), saltNonce, getChainId()));
+        proxy = deployProxy(_singleton, salt);
         if (initializer.length > 0)
```

Comment:
I realized that we could move this to the `deployProxy` method or use the `createProxyWithNonce` method in this function (with the `saltNonce` being `keccak256(abi.encodePacked(saltNonce, callback))` (as it is for [createProxyWithCallback](https://github.com/safe-global/safe-contracts/pull/468/files#diff-7a756ca403cd3005982b1a2b902059a02ed81661eae64998b42c8d51972ea5f0R94)). This would reduce code duplication.

Reply:
good call, please check it now. I had to import the runtime bytecode from solc artifacts to be used in tests though.